File: 백준/Gold/2623. 음악프로그램/음악프로그램.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.Buffer;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int n, m, cnt;
    static int[] ind = new int[1004];
    static ArrayList<Integer>[] adj;
    static ArrayList<Integer> ret = new ArrayList<>();
    static Queue<Integer> q = new ArrayDeque<>();

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        // 인접 리스트
        adj = new ArrayList[n + 4]; // 리스트
        for (int i = 0; i < n + 4; i++) { // 리스트 안에 리스트
            adj[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            int prev = -1;

            st = new StringTokenizer(br.readLine());
            // 3 1 4 3
            int k = Integer.parseInt(st.nextToken()); // 3

            for(int j = 0; j < k; j++){// 1 4 3
                int temp = Integer.parseInt(st.nextToken());// 1

                if(j != 0){ //j != 0
                    adj[prev].add(temp);
                    ind[temp]++;
                    prev = temp;
                }else{ // j == 0
                    prev = temp; // prev = 1
                }
            }
        }

        for (int i = 1; i <= n; i++) {
            if (ind[i] == 0) q.add(i); // indegree == 0거 q에 넣기
        }

        while (!q.isEmpty()) {
            int nowNode = q.poll(); // front, pop
//            System.out.println(nowNode);
            ret.add(nowNode);

            for (int nextNode : adj[nowNode]) {
                ind[nextNode]--; // indegree--

                if (ind[nextNode] == 0) {
                    q.add(nextNode);// indegree == 0거 q에 넣기
                }
            }
        }

        if(ret.size() != n || n == cnt){
            System.out.println(0);
        }else{
            for(int i: ret)
                System.out.println(i);
        }

    }
}

File: 백준/Gold/2623. 음악프로그램/음악프로그램.cc
#include <bits/stdc++.h>
using namespace std;
int n, m;
vector<int> adj[1004];
int ind[1004];
queue<int> q;
vector<int> ret;
int main(){
    ios_base::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);

    cin  >> n >> m;
    
    for(int i = 0; i < m; i++){
        int k;
        cin >> k;

        int prev = 0;
        for(int j = 0; j < k; j++){
            int temp;
            cin >> temp;
            if(prev != 0){
                int f = prev, t = temp;

                adj[f].push_back(t);
                ind[temp]++;
            }
            prev = temp;
        }
    }

    for(int i = 1; i <= n; i++){ // 1 <= X <= n
        if(ind[i] == 0)q.push(i); // indegree 0인 노드 -> queue에 넣기
    }

    while(q.size()){
        int nowNode = q.front();
        q.pop();
    
        ret.push_back(nowNode);

        for(int nextNode: adj[nowNode]){
            ind[nextNode]--;

            if (ind[nextNode] == 0) {
               q.push(nextNode); // indegree 0인 노드 -> queue에 넣기
            }
        }
    }

    if(ret.size() == n){
        for(int i: ret){
            cout << i << "\n";
        }
    }else{
        cout << 0 << "\n";
    }

    return 0;
}